src/org/omegat/core/dictionaries/LingvoDSL.java

```diff
@@ -64,161 +52,226 @@
  */
 public class LingvoDSL implements IDictionaryFactory {
 
-    // An ordered list of Pair of Regex pattern and replacement string
-    private static final TreeMap<Pattern, String> TAG_REPLACEMENTS = new TreeMap<>(
-            Comparator.comparing(Pattern::pattern));
-
-    // A sentinel that is replaced with the absolute path of the dictionary in
-    // {@link #loadData}
-    private static final String DICT_DIR_TAG = "@dir@";
-    private static final int BLOCKSIZE = 8192;
-
     @Override
     public final boolean isSupportedFile(File file) {
         return file.getPath().endsWith(".dsl") || file.getPath().endsWith(".dsl.dz");
     }
 
     @Override
     public final IDictionary loadDict(File file) throws Exception {
-        return loadDict(file, new Language(Locale.getDefault()));
+        return new LingvoDSLDict(file);
     }
 
-    @Override
-    public final IDictionary loadDict(File file, Language language) throws Exception {
-        return new LingvoDSLDict(file, language);
+    static class LingvoDSLDict implements IDictionary {
+        protected final DslDictionary data;
+        private final HtmlVisitor htmlVisitor;
+
+        LingvoDSLDict(File file) throws Exception {
+            data = DslDictionary.loadDictionary(file);
+            htmlVisitor = new HtmlVisitor(file.getParent());
+        }
+
+        @Override
+        public List<DictionaryEntry> readArticles(String word) {
+            return data.lookup(word).getEntries(htmlVisitor).stream()
+                    .map(e -> new DictionaryEntry(e.getKey(), e.getValue()))
```

Comment:
Here is a place we can see how to use the parser.

- `DslDictionary#loadDictionary` is a builder method to load `dsl` file.
- `DslDictionary#lookup` returns `DslResult` object that hold result articles' AST.
- `DslResult#getEntries` returns `List<Entry<String, T>>` when pass `visitor<T>` object.

Reply:
Thanks.

Two minor thoughts:

1. As I understand it, Streams are significantly less efficient (make much more garbage) than simple `for` loops for this kind of trivial transformation. You might consider switching to a `for` loop if you're at all worried about that.
2. The only difference between `readArticles` and `readArticlesPredictive` is the method called on `this.data`. You could consolidate the transformation to `List<DictionaryEntry>`.

Reply:
Thank you for the feedback. I'd like to try both ideas.